File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/FlareAuthShell.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/**
 * Product authentication shell shared by Flare Compose applications.
 *
 * The kit owns the responsive brand/form composition. Hosts provide localized
 * copy and business controls through [content].
 */
@Composable
fun FlareAuthShell(
    product: String,
    title: String,
    subtitle: String,
    modifier: Modifier = Modifier,
    tagline: String? = null,
    headline: String? = null,
    description: String? = null,
    backLabel: String? = null,
    onBack: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    FlareScreen(surface = FlareScreenSurface.Brand, scroll = false) {
        BoxWithConstraints(modifier.fillMaxSize()) {
            val wide = maxWidth >= FlareSizes.appShellCompactMinWidth
            Box(Modifier.fillMaxSize().background(flareColors().bgPrimary)) {
                if (wide) {
                    Row(
                        Modifier
                            .fillMaxSize()
                            .background(flareColors().bgPrimary),
                    ) {
                        BrandPanel(
                            product = product,
                            tagline = tagline,
                            headline = headline,
                            description = description,
                            modifier = Modifier.weight(5f).fillMaxHeight(),
                        )
                        FormPanel(
                            product = product,
                            tagline = tagline,
                            title = title,
                            subtitle = subtitle,
                            backLabel = backLabel,
                            onBack = onBack,
                            showBrand = false,
                            modifier = Modifier.weight(6f).fillMaxHeight(),
                            content = content,
                        )
                    }
                } else {
                    FormPanel(
                        product = product,
                        tagline = tagline,
                        title = title,
                        subtitle = subtitle,
                        backLabel = backLabel,
                        onBack = onBack,
                        showBrand = true,
                        modifier = Modifier.fillMaxSize().background(flareColors().bgPrimary),
                        content = content,
                    )
                }
            }
        }
    }
}

@Composable
private fun BrandPanel(
    product: String,
    tagline: String?,
    headline: String?,
    description: String?,
    modifier: Modifier = Modifier,
) {
    val colors = flareColors()
    Column(
        modifier.background(
            Brush.linearGradient(
                listOf(lerp(colors.info, colors.bgPrimary, 0.80f), lerp(colors.primary, colors.bgSecondary, 0.90f)),
            ),
        ).padding(56.dp),
        verticalArrangement = Arrangement.Top,
    ) {
        BrandLockup(product, tagline, 72.dp)
        Box(Modifier.weight(1f).fillMaxWidth()) {
            Column(Modifier.align(Alignment.CenterStart).widthIn(max = 420.dp)) {
                headline?.let {
                    Text(
                        it,
                        color = colors.textPrimary,
                        fontSize = FlareSizes.fontSize5xl,
                        fontWeight = FontWeight.Bold,
                        lineHeight = FlareSizes.fontSize5xl * FlareSizes.lineHeightTight,
                    )
                }
                description?.let {
                    Text(
                        it,
                        modifier = Modifier.padding(top = FlareSizes.spacingLg),
                        color = colors.textSecondary,
                        fontSize = FlareSizes.fontSize2xl,
                        lineHeight = FlareSizes.fontSize2xl * FlareSizes.lineHeightRelaxed,
                    )
                }
            }
        }
    }
}

@Composable
private fun FormPanel(
    product: String,
    tagline: String?,
    title: String,
    subtitle: String,
    backLabel: String?,
    onBack: (() -> Unit)?,
    showBrand: Boolean,
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = flareColors()
    Box(modifier.background(colors.bgPrimary).verticalScroll(rememberScrollState())) {
        Column(
            Modifier
                .align(if (showBrand) Alignment.TopCenter else Alignment.Center)
                .widthIn(max = 500.dp)
                .fillMaxWidth()
                .padding(horizontal = if (showBrand) FlareSizes.spacing2xl else 64.dp, vertical = if (showBrand) FlareSizes.spacing2xl else 48.dp),
        ) {
            if (showBrand) {
                BrandLockup(product, tagline, 52.dp)
                Spacer(Modifier.size(FlareSizes.spacing2xl))
            }
            if (backLabel != null && onBack != null) {
                TextButton(onClick = onBack) {
                    Icon(Icons.AutoMirrored.Outlined.ArrowBack, contentDescription = null, modifier = Modifier.size(FlareSizes.iconSizeSm))
                    Spacer(Modifier.size(FlareSizes.spacingSm))
                    Text(backLabel)
                }
            }
            Text(
                title,
                modifier = Modifier.semantics { heading() },
                color = colors.textPrimary,
                fontSize = FlareSizes.fontSize5xl,
                fontWeight = FontWeight.Bold,
            )
            Text(
                subtitle,
                modifier = Modifier.padding(top = FlareSizes.spacingSm),
                color = colors.textSecondary,
                fontSize = FlareSizes.fontSize2xl,
                lineHeight = FlareSizes.fontSize2xl * FlareSizes.lineHeightRelaxed,
            )
            Spacer(Modifier.size(FlareSizes.spacing2xl))
            content()
        }
    }
}

@Composable
private fun BrandLockup(product: String, tagline: String?, logoSize: androidx.compose.ui.unit.Dp) {
    val colors = flareColors()
    Row(verticalAlignment = Alignment.CenterVertically) {
        FlareBrandLogo(size = logoSize)
        Column(Modifier.padding(start = FlareSizes.spacingLg)) {
            Text(product, color = colors.textPrimary, fontSize = FlareSizes.fontSize3xl, fontWeight = FontWeight.Bold)
            tagline?.let {
                Text(
                    it,
                    modifier = Modifier.padding(top = FlareSizes.spacingXs),
                    color = colors.primaryText,
                    fontSize = FlareSizes.fontSizeSm,
                    fontWeight = FontWeight.SemiBold,
                    letterSpacing = FlareSizes.fontSize2xs * 0.05f,
                )
            }
        }
    }
}

File: packages/flutter-im-ui/lib/src/components/flare_auth_shell.dart
import 'package:flutter/material.dart';

import '../tokens/flare_tokens.dart';
import 'flare_brand_logo.dart';
import 'flare_screen.dart';

/// Product-grade authentication scaffold shared by Flare applications.
///
/// The kit owns the responsive brand/form composition and semantic hierarchy;
/// hosts provide localized copy and their business form as [child].
class FlareAuthShell extends StatelessWidget {
  const FlareAuthShell({
    super.key,
    required this.product,
    required this.title,
    required this.subtitle,
    required this.child,
    this.tagline,
    this.headline,
    this.description,
    this.backLabel,
    this.onBack,
  });

  final String product;
  final String title;
  final String subtitle;
  final String? tagline;
  final String? headline;
  final String? description;
  final String? backLabel;
  final VoidCallback? onBack;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    final colors = FlareColors.of(context);
    return LayoutBuilder(
      builder: (context, viewport) {
        final wide = viewport.maxWidth >= FlareSizes.appShellCompactMinWidth;
        return FlareScreen(
          surface: wide ? FlareScreenSurface.brand : FlareScreenSurface.surface,
          scroll: false,
          child: ColoredBox(
            color: colors.bgPrimary,
            child: wide
                ? Row(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      Expanded(flex: 5, child: _BrandPanel(this)),
                      Expanded(
                        flex: 6,
                        child: _FormPanel(this, showBrand: false),
                      ),
                    ],
                  )
                : _FormPanel(this, showBrand: true),
          ),
        );
      },
    );
  }
}

class _BrandPanel extends StatelessWidget {
  const _BrandPanel(this.shell);
  final FlareAuthShell shell;

  @override
  Widget build(BuildContext context) {
    final colors = FlareColors.of(context);
    return DecoratedBox(
      decoration: BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topLeft,
          end: Alignment.bottomRight,
          colors: [
            Color.alphaBlend(
              colors.info.withValues(alpha: 0.18),
              colors.bgPrimary,
            ),
            Color.alphaBlend(
              colors.primary.withValues(alpha: 0.10),
              colors.bgSecondary,
            ),
          ],
        ),
      ),
      child: Padding(
        padding: const EdgeInsets.all(56),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            _BrandLockup(shell, logoSize: 72),
            Expanded(
              child: Align(
                alignment: Alignment.centerLeft,
                child: ConstrainedBox(
                  constraints: const BoxConstraints(maxWidth: 420),
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      if (shell.headline case final headline?)
                        Text(
                          headline,
                          style: TextStyle(
                            color: colors.textPrimary,
                            fontSize: FlareSizes.fontSize5xl,
                            fontWeight: FontWeight.w700,
                            height: FlareSizes.lineHeightTight,
                          ),
                        ),
                      if (shell.description case final description?) ...[
                        const SizedBox(height: FlareSizes.spacingLg),
                        Text(
                          description,
                          style: TextStyle(
                            color: colors.textSecondary,
                            fontSize: FlareSizes.fontSize2xl,
                            height: FlareSizes.lineHeightRelaxed,
                          ),
                        ),
                      ],
                    ],
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _FormPanel extends StatelessWidget {
  const _FormPanel(this.shell, {required this.showBrand});
  final FlareAuthShell shell;
  final bool showBrand;

  @override
  Widget build(BuildContext context) {
    final colors = FlareColors.of(context);
    final horizontal = showBrand ? FlareSizes.spacing2xl : 64.0;
    final vertical = showBrand ? FlareSizes.spacing2xl : 48.0;
    return ColoredBox(
      color: colors.bgPrimary,
      child: LayoutBuilder(
        builder: (context, constraints) => SingleChildScrollView(
          padding: EdgeInsets.symmetric(
            horizontal: horizontal,
            vertical: vertical,
          ),
          child: ConstrainedBox(
            constraints: BoxConstraints(
              minHeight: (constraints.maxHeight - vertical * 2)
                  .clamp(0.0, double.infinity)
                  .toDouble(),
            ),
            child: Align(
              alignment: showBrand ? Alignment.topCenter : Alignment.center,
              child: ConstrainedBox(
                constraints: const BoxConstraints(maxWidth: 500),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    if (showBrand) ...[
                      _BrandLockup(shell, logoSize: 52),
                      const SizedBox(height: FlareSizes.spacing2xl),
                    ],
                    if (shell.onBack != null && shell.backLabel != null)
                      Align(
                        alignment: AlignmentDirectional.centerStart,
                        child: TextButton.icon(
                          onPressed: shell.onBack,
                          icon: const Icon(
                            Icons.arrow_back,
                            size: FlareSizes.iconSizeSm,
                          ),
                          label: Text(shell.backLabel!),
                          style: TextButton.styleFrom(
                            foregroundColor: colors.textLink,
                            minimumSize: const Size(
                              FlareSizes.touchTargetMin,
                              FlareSizes.touchTargetMin,
                            ),
                          ),
                        ),
                      ),
                    Semantics(
                      header: true,
                      child: Text(
                        shell.title,
                        style: TextStyle(
                          color: colors.textPrimary,
                          fontSize: FlareSizes.fontSize5xl,
                          fontWeight: FontWeight.w700,
                          height: FlareSizes.lineHeightTight,
                        ),
                      ),
                    ),
                    const SizedBox(height: FlareSizes.spacingSm),
                    Text(
                      shell.subtitle,
                      style: TextStyle(
                        color: colors.textSecondary,
                        fontSize: FlareSizes.fontSize2xl,
                        height: FlareSizes.lineHeightRelaxed,
                      ),
                    ),
                    const SizedBox(height: FlareSizes.spacing2xl),
                    shell.child,
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class _BrandLockup extends StatelessWidget {
  const _BrandLockup(this.shell, {required this.logoSize});
  final FlareAuthShell shell;
  final double logoSize;

  @override
  Widget build(BuildContext context) {
    final colors = FlareColors.of(context);
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        FlareBrandLogo(size: logoSize),
        const SizedBox(width: FlareSizes.spacingLg),
        Flexible(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                shell.product,
                style: TextStyle(
                  color: colors.textPrimary,
                  fontSize: FlareSizes.fontSize3xl,
                  fontWeight: FontWeight.w700,
                ),
              ),
              if (shell.tagline case final tagline?) ...[
                const SizedBox(height: FlareSizes.spacingXs),
                Text(
                  tagline,
                  style: TextStyle(
                    color: colors.primaryText,
                    fontSize: FlareSizes.fontSizeSm,
                    fontWeight: FontWeight.w600,
                    letterSpacing: 0.5,
                  ),
                ),
              ],
            ],
          ),
        ),
      ],
    );
  }
}
